### ISSUE TITLE
Zce gcc, binutils

### ISSUE THREAD

**Issue by jjscheel** [riscv-admin/dev-partners#5]
### Technical Group

Code Size Reduction TG

### ratification-pkg

Code Size

### Technical Liaison

Tariq Kurd

### Task Category

Compilers/Toolchain

### Task Sub Category

- [X] gcc
- [X] binutils
- [ ] gdb
- [ ] intrinsics
- [ ] Java
- [ ] KVM
- [ ] ld
- [ ] llvm
- [ ] Linux kernel
- [ ] QEMU
- [ ] Spike

### Ratification Target

3Q2023

### Statement of Work (SOW)

[SOW link](https://docs.google.com/document/d/1k0xdoMMUpNZ6f4vGy0B0_66RYecToKEoIoSc9P59chA/)

**SOW Signoffs:** (delete those not needed)
  - Task group liaison sign-off date:
  - Development partner sign-off date:
  - Toolchains & Runtimes sign-off date (if gcc, LLVM, binutils, gdb work):

### Waiver

- [ ] Freeze
- [ ] Ratification

**Comment by jjscheel** [comment]
Patches remain in review per 3/28 meeting.  No action needed.

**Comment by jjscheel** [comment]
Ok, @pz9115.  Yes, I understand that Kito is very busy with RVV Intrinsic work!

**Comment by pz9115** [comment]
Kito plan to finish this in gcc 14:

https://gcc.gnu.org/pipermail/gcc-patches/2023-April/616266.html
https://gcc.gnu.org/pipermail/gcc-patches/2023-April/616267.html

**Comment by kito-cheng** [comment]
@pz9115 let you know I'll take a close review this week.

**Comment by pz9115** [comment]
I got serveral suggestions for the gcc patches and now discussing in zcmp extension implement.

**Comment by jjscheel** [comment]
Thanks, @pz9115!

**Comment by pz9115** [comment]
Resend gcc patch and binutils patch to upstream, Eswin help optimal the zcmp extension implement. Plan to sync two week later :)

**Comment by pz9115** [comment]
Zca/f/d/b extensions patch were merged, zcmp/t patches are preparing.

**Comment by pz9115** [comment]
You can find a total implement version in  this issue: https://github.com/riscv-collab/riscv-gnu-toolchain/issues/1288

**Comment by Micronana** [comment]
Hi，Pz9115
I just compile the gcc toolchain which support Zc* as you mentioned there:  https://github.com/riscv-collab/riscv-gnu-toolchain/issues/1288.  
As Zc is used to reduce code size,  but when I use this Zc* gcc to compile coremark with   rv32imc,  the code size is 2% larger than than the normal gcc in master branch.    So even I compile the program with rv32im_cmt,  the code size is still larger than normal gcc with rv32imc option. 
In summary,  the Zc toolchain I tried has not reduce the code size.  Is it  expected?

**Comment by pz9115** [comment]
Sever Binutils patch merged into upstream, Zcmp/Zcmt is still work in progress.

**Comment by jjscheel** [comment]
@pz9115, thanks for the update.  Can we now claim that all Zc binutils work is upstream?

**Comment by jjscheel** [comment]
@pz9115, what progress did we make on Zcmp/Zcmt patch merge?

**Comment by pz9115** [comment]
We still work on Zcmp/t implement in Binutils, and some gcc Zcmp patches merged into upstream in past two weeks.

**Comment by pz9115** [comment]
WIP，will resend binutils patches recently.

**Comment by pz9115** [comment]
Discussed in the last RISC-V GNU toolchain meeting, WIP.

**Comment by pz9115** [comment]
Resent patches to upstream, status change to in review
https://sourceware.org/pipermail/binutils/2023-November/130690.html
https://sourceware.org/pipermail/binutils/2023-November/130691.html

**Comment by pz9115** [comment]
Also updated the Zcmt extension patch:
https://sourceware.org/pipermail/binutils/2023-November/130849.html

**Comment by pz9115** [comment]
Waiting for review yet.

**Comment by pz9115** [comment]
Receive Kito's review comments and now modify the patch, prepare to resend them within this week.

**Comment by pz9115** [comment]
Receive more comments from Kito to option the patch:

https://sourceware.org/pipermail/binutils/2024-January/131575.html

https://sourceware.org/pipermail/binutils/2024-January/thread.html

Resend fifth version of patch, keep reviewing:

ZCMP: https://sourceware.org/pipermail/binutils/2024-January/132237.html

**Comment by pz9115** [comment]
Received new comments from Kito, WIP.

**Comment by pz9115** [comment]
Will resend new patches this week.

**Comment by pz9115** [comment]
Zcmp patch merged into Binutils upstream, now working on Zcmt part.

https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=9132c8152b899a1683bc886f8ba76bedadb48aa1

**Comment by jjscheel** [comment]
Nice work, @pz9115!  I've added your patch to the description above.  Is Zcmt the only other patch needed here?

**Comment by pz9115** [comment]
Now we only remain Zcmt patch, it needs some linker part changes, still WIP.

**Comment by pz9115** [comment]
WIP.

**Comment by jjscheel** [comment]
How much longer before we have a submitted PR for Zcmt, @pz9115?

**Comment by pz9115** [comment]
> How much longer before we have a submitted PR for Zcmt, @pz9115?

I think it will take about two weeks, I will finish this job within this month.

**Comment by pz9115** [comment]
Yes, it is, still wait for reviewing.

**Comment by pz9115** [comment]
Now new updates in these two weeks.

**Comment by pz9115** [comment]
No review received.

**Comment by pz9115** [comment]
Keep waiting...

**Comment by pz9115** [comment]
No new conditions here.

**Comment by pz9115** [comment]
During the test, my colleague found that two instructions of zcmp were missing. I resubmitted the patch of cm.mv instruction, which has been merged into binutils upstream.

https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=ca2590d7804b4ea563eec6f1127ed17a00c30315

Rebasing zcmt patch into binutils 2.43

**Comment by pz9115** [comment]
Received Nelson's helpful comments, preparing a new patch.

**Comment by pz9115** [comment]
Meet some conflict issue, still prepare new patch.

**Comment by pz9115** [comment]
Rebased and sent to the mailing list, waiting for review.
https://sourceware.org/pipermail/binutils/2024-October/137437.html

**Comment by pz9115** [comment]
Still waiting for review.

**Comment by pz9115** [comment]
Zcmt merged into binutils upstream.
https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=00ef37e860d505f41a621708c62f3e9363868ee1

**Comment by jjscheel** [comment]
@tariqkurd-repo, we appear to have all gcc and binutils patches upstream.  Would you kindly do a quick check and let me know if you agree with considering this SOW complete?  I'm sure @pz9115 can answer any questions.

**Comment by tariqkurd-repo** [comment]
Sorry Jeff I missed the message 
I trust you to review whether the SoW is complete, it's a long time since I've looked at any of the work related to this.

**Comment by pz9115** [comment]
Using new approach to emliate table-jump relaxation support in both ld/lld side. Waiting for review.

ld: https://sourceware.org/pipermail/binutils/2026-June/149842.html
lld: https://github.com/llvm/llvm-project/pull/206338
